src/jrd/exe.cpp

```diff
@@ -917,10 +920,140 @@ void EXE_activate(thread_db* tdbb, Request* request, jrd_tra* transaction)
 }
 
 
+// Execute function. A shortcut for node-based function but required for external functions.
+void EXE_execute_function(thread_db* tdbb, Request* request, jrd_tra* transaction,
+	ULONG inMsgLength, UCHAR* inMsg, ULONG outMsgLength, UCHAR* outMsg)
+{
+	if (const auto function = request->getStatement()->function; function && function->fun_external)
+	{
+		activate_request(tdbb, request, transaction);
+
+		const auto attachment = tdbb->getAttachment();
+
+		// Ensure the cancellation lock can be triggered
+		const auto lock = attachment->att_cancel_lock;
+		if (lock && lock->lck_logical == LCK_none)
+			LCK_lock(tdbb, lock, LCK_SR, LCK_WAIT);
+
+		const SavNumber savNumber = savepointStart(request, transaction);
+
+		if (!request->req_transaction)
+			ERR_post(Arg::Gds(isc_req_no_trans));
+
+		try
+		{
+			// Save the old pool and request to restore on exit
+			StmtNode::ExeState exeState(tdbb, request, request->req_transaction);
+			Jrd::ContextPoolHolder context(tdbb, request->req_pool);
+
+			fb_assert(!request->req_caller);
+			request->req_caller = exeState.oldRequest;
+
+			tdbb->tdbb_flags &= ~(TDBB_stack_trace_done | TDBB_sys_error);
+
+			// Execute stuff until we drop
+
+			const auto profilerManager = attachment->isProfilerActive() && !request->hasInternalStatement() ?
+				attachment->getProfilerManager(tdbb) : nullptr;
+			SINT64 profilerInitialTicks = profilerManager ? profilerManager->queryTicks() : 0;
+
+			try
+			{
+				JRD_reschedule(tdbb);
+
+				function->fun_external->execute(tdbb, request, transaction, inMsgLength, inMsg, outMsgLength, outMsg);
+			}
```

Comment:
Even if `att_cancel_lock` is active and cancellation request has been posted to our attachment, the external function cannot be cancelled during its execution. But maybe we should call `tdbb->checkCancelState()` immediately after its execution to react quickly? And maybe `JRD_reschedule()` will be not needed after that change, given that we always check out the engine during function invocation.

Reply:
Please see if it's now correct.